server/Main.hs

```diff
@@ -234,7 +220,9 @@ completeLogin sessions users = do
     verifyLogin :: SessionId -> Session -> Fido2.UserId -> Fido2.Challenge -> Scotty.ActionM ()
     verifyLogin sessionId session userId challenge = do
       credential <- Scotty.jsonData @(Fido2.PublicKeyCredential Fido2.AuthenticatorAssertionResponse)
-      credentials <- (liftIO $ getUserCredentials userId users) >>= handleError
+      credentials <- liftIO $ Database.withTransaction db $ \tx -> do
```

Comment:
We already have the credential here, so we can look up only that one by id, right?

Reply:
That would require us to trust that the user actually respected the list of `allowedCredentials` that we passed to them. I'm not sure whether that is safe (but I suspect it isn't).

This code ensures that we only check the challenge against the list of credentials that are actually associated with the current user in the DB.

We can also change this to a lookup by both the user ID and the credential ID of course, but I'm not sure if there is a reason for choosing one over the other.

Reply:
> We can also change this to a lookup by both the user ID and the credential ID of course

Yes let's do this.